olive/engine.py

```diff
@@ -134,38 +135,47 @@ def register(
         Register a pass
         """
         if name is not None:
-            assert name not in self._passes, f"Pass with name {name} already registered"
+            assert name not in self.passes, f"Pass with name {name} already registered"
         else:
             id = 0
             while True:
                 name = p.__class__.__name__
                 if id > 0:
                     name = f"{name}_{id}"
-                if name not in self._passes:
+                id += 1
+                if name not in self.passes:
                     break
 
-        self._passes[name] = p
-        self._pass_order.append(name)
-        self.hosts[name] = host
-        self._evaluators[name] = evaluator
-        self._clean_pass_run_cache[name] = clean_run_cache
+        if self.search_strategy is None and len(p.search_space()) > 0:
+            raise ValueError(f"Search strategy is None but pass {name} has search space")
 
-    def run(self, input_model: OliveModel, verbose: bool = False):
+        self.passes[name] = {
+            "pass": p,
+            "host": host,
+            "evaluator": evaluator,
+            "clean_run_cache": clean_run_cache,
+        }
+        self.pass_order.append(name)
+
+    def run(self, input_model: OliveModel, verbose: bool = False, output_dir: str = None, output_name: str = None):
         """
         Run all the registered Olive passes on the input model and produce one or more candidate models.
         """
         if not self._initialized:
             self.initialize()
 
+        if self.search_strategy is None:
+            return self._run_no_search(input_model, verbose, output_dir, output_name)
+
```

Comment:
Is it possible to avoid separate run function?
One way to look at this path is -- The search space has only single search point for each Pass.

Reply:
Yes, we could combine them. We can create a dummy search strategy to run the passes without search using the pre-existing logic. This also addresses a part of @trajepl 's comment about splitting the logic.

Reply:
Done